File: src/views/applets/MazeGenerator/MazeGenerator.ts

import { filter, map } from 'lodash';
import Alea from 'alea'

export enum ECellCategory {
    Empty,
    Default,
    Shortcut,
    Loop,
    Start,
    End,
}

export enum ECellAccessibilityLevel {
    Crossroad,
    Fork,
    Transitive,
    DeadEnd,
    Isolated,
}

export enum EGenerationOrder {
    Pop,
    Shift
}

export class MazeCell {
    constructor(maze_generator: MazeGenerator, x: number, y: number, category: ECellCategory = ECellCategory.Empty) {
        this.maze_generator = maze_generator;
        this.walls = { north: true, east: true, south: true, west: true }
        this.x = x;
        this.y = y;
        this.visited = false;
        this.category = category;
    }

    // PROPS

    public index: number
    public distance?: number;
    public x: number;
    public y: number;
    public walls: { north: boolean, east: boolean, south: boolean, west: boolean } = null;
    public category: ECellCategory;
    public visited: boolean;

    protected maze_generator?: MazeGenerator;
    protected neighbours_offsets = [
        { x: -1, y: 0 },
        { x: 0, y: -1 },
        { x: 1, y: 0 },
        { x: 0, y: 1 },
    ]

    public get walls_count(): number {
        let result = 0;
        if (this.walls.north) result++;
        if (this.walls.east) result++;
        if (this.walls.south) result++;
        if (this.walls.west) result++;
        return result;
    }

    // METHODS
    public get_closed_neighbours(): MazeCell[] {
        return filter(this.get_all_neighbours(), (cell: MazeCell) => {
            return this.get_wall_between(cell) && cell.walls_count !== ECellAccessibilityLevel.Isolated
        })
    }
    public get_open_neighbours(): MazeCell[] {
        return filter(this.get_all_neighbours(), (cell) => {
            return !this.get_wall_between(cell)
        })
    }
    public get_all_neighbours(): MazeCell[] {
        let cells = this.maze_generator.cells;
        return filter(
            map(this.neighbours_offsets, (offset, index) => { return this.get_neighbour_cell(index) }),
            (cell) => { return cell !== null }
        )
    }
    public remove_wall_between(cell2: MazeCell): void {
        this.set_wall_between(cell2, false);
    }
    public add_wall_between(cell2: MazeCell): void {
        this.set_wall_between(cell2, true);
    }
    public set_wall_between(cell2: MazeCell, value: boolean): void {
        let cell1 = this
        let x = cell1.x - cell2.x;
        let y = cell1.y - cell2.y;

        if (x === 1) {
            cell1.walls.west = value;
            cell2.walls.east = value;
        } else if (x === -1) {
            cell1.walls.east = value;
            cell2.walls.west = value;
        }

        if (y === 1) {
            cell1.walls.north = value;
            cell2.walls.south = value;
        } else if (y === -1) {
            cell1.walls.south = value;
            cell2.walls.north = value;
        }
    }
    public get_wall_between(cell2: MazeCell): boolean {
        let x = this.x - cell2.x;
        let y = this.y - cell2.y;

        if (x === 1) {
            return this.walls.west;
        } else if (x === -1) {
            return this.walls.east;
        }

        if (y === 1) {
            return this.walls.north;
        } else if (y === -1) {
            return this.walls.south;
        }
    }
    protected get_neighbour_cell(offset: number) {
        if (offset < 0 || offset > 3) {
            throw new Error("offset must be between 0 and 3");
        }
        let x = this.x + this.neighbours_offsets[offset].x
        let y = this.y + this.neighbours_offsets[offset].y

        if (x < 0 || x >= this.maze_generator.grid_size || y < 0 || y >= this.maze_generator.grid_size) {
            return null
        }

        return this.maze_generator.cells[x][y]
    }
}

export class MazeGenerator {
    constructor() {
        this.cells = []
    }

    //  PROPS
    public cells: MazeCell[][]
    public start_cell: MazeCell
    public end_cell: MazeCell

    public seed: number = 0
    public grid_size: number = 4
    public sparseness: number = 0.5
    public dead_ends_ratio: number = 0.75
    public shortcuts_ratio: number = 0.2
    public generation_order: EGenerationOrder = EGenerationOrder.Shift

    protected current_cell_index: number = 0
    protected seeded_random

    public get max_cells_count() {
        return this.grid_size * this.grid_size;
    }
    // METHODS

    protected update_generator() {
        this.seeded_random = Alea(this.seed.toString())
    }
    protected get_random_cell(): MazeCell {
        return this.cells[Math.floor(this.seeded_random() * this.grid_size)][Math.floor(this.seeded_random() * this.grid_size)];
    }
    protected initialize() {
        this.update_generator();
        this.current_cell_index = 0;
        this.cells = [];

        for (let i = 0; i < this.grid_size; i++) {
            this.cells[i] = [];
            for (let j = 0; j < this.grid_size; j++) {
                this.cells[i][j] = new MazeCell(this, i, j, ECellCategory.Empty)
            }
        }
    }
    public generate() {
        this.initialize();

        let stack = [];
        let current_cell = this.start_cell = this.get_random_cell();
        let end_cell: MazeCell = current_cell;
        current_cell.distance = 0;
        current_cell.index = this.current_cell_index;
        current_cell.visited = true;
        current_cell.category = ECellCategory.Start;
        stack.push(current_cell);

        while (stack.length > 0) {
            if (current_cell.index >= Math.max(1, (1 - this.sparseness) * this.max_cells_count)) {
                break;
            }

            let neighbours = current_cell.get_all_neighbours();
            let unvisited_neighbours = neighbours.filter(cell => cell.visited === false);

            if (unvisited_neighbours.length > 0) {
                let random_neighbour = unvisited_neighbours[Math.floor(this.seeded_random() * unvisited_neighbours.length)];
                current_cell.remove_wall_between(random_neighbour);
                random_neighbour.visited = true;
                random_neighbour.distance = current_cell.distance + 1;
                random_neighbour.index = this.current_cell_index++;
                stack.push(random_neighbour);
                current_cell = random_neighbour;
                end_cell = current_cell.distance > end_cell.distance ? current_cell : end_cell;
            } else {
                switch (this.generation_order) {
                    case EGenerationOrder.Pop: {
                        current_cell = stack.pop();
                        break;
                    }
                    case EGenerationOrder.Shift: {
                        current_cell = stack.shift();
                        break;
                    }
                }

            }
        }

        end_cell.category = ECellCategory.End;
        this.end_cell = end_cell;

        // DEAD ENDS
        let dead_ends = this.find_cells_with_accessibility(ECellAccessibilityLevel.DeadEnd);
        dead_ends.forEach(cell => {
            if (this.seeded_random() > (this.dead_ends_ratio)) {
                if (cell.category === ECellCategory.Start || cell.category === ECellCategory.End) {
                    return
                }

                let closed_neighbours = cell.get_closed_neighbours(this.cells);

                if (closed_neighbours.length > 0) {
                    let random_neighbour = closed_neighbours[Math.floor(this.seeded_random() * closed_neighbours.length)];

                    cell.remove_wall_between(random_neighbour);
                    cell.category = ECellCategory.Loop;
                }
            }
        })

        // SHORTCUTS
        let transitive_cells = this.find_cells_with_accessibility(ECellAccessibilityLevel.Transitive);
        transitive_cells.forEach(cell => {
            let closed_neighbours = cell.get_closed_neighbours();

            if (closed_neighbours.length > 1) {
                let random_neighbour = closed_neighbours[Math.floor(this.seeded_random() * closed_neighbours.length)];
                if (this.seeded_random() < this.shortcuts_ratio) {
                    cell.remove_wall_between(random_neighbour);
                    cell.category = ECellCategory.Shortcut;
                }
            }
        })

        this.for_each_cell((cell) => {
            if (cell.walls_count === ECellAccessibilityLevel.Isolated) {
                cell.category = ECellCategory.Empty;
            }
        })

        this.reset_visited()
    }
    find_cells_with_accessibility(level: ECellAccessibilityLevel) {
        let result = [];
        this.for_each_cell((cell) => {
            if (cell.walls_count === level) {
                result.push(cell);
            }
        })

        return result;
    }

    public reset_visited() {
        this.for_each_cell((cell) => {
            cell.visited = false;
        })
    }

    protected for_each_cell(callback: (cell: MazeCell) => void) {
        this.cells.forEach(cells => {
            cells.forEach(cell => {
                callback(cell);
            })
        })
    }
}

File: src/views/applets/MazeGenerator/variations/maze_generator.py
import json
import random
from enum import Enum
from typing import Optional, List, Dict, Tuple

class ECellCategory(Enum):
    Empty = 0
    Default = 1
    Shortcut = 2
    Loop = 3
    Start = 4
    End = 5

class ECellAccessibilityLevel(Enum):
    Crossroad = 0
    Fork = 1
    Transitive = 2
    DeadEnd = 3
    Isolated = 4

class EGenerationOrder(Enum):
    Pop = 0
    Shift = 1

class MazeCell:
    def __init__(self, maze_generator, x: int, y: int, category: ECellCategory = ECellCategory.Default):
        self.maze_generator = maze_generator
        self.x = x
        self.y = y
        self.walls = {'north': True, 'east': True, 'south': True, 'west': True}
        self.visited = False
        self.category = category
        self.route = None
        self.distance = None
        self.neighbours_offsets = [(-1, 0), (0, -1), (1, 0), (0, 1)]

    @property
    def walls_count(self):
        return sum(self.walls.values())

    def get_closed_neighbours(self):
        return [cell for cell in self.get_all_neighbours() if self.get_wall_between(cell) and cell.walls_count != ECellAccessibilityLevel.Isolated]

    def get_open_neighbours(self):
        return [cell for cell in self.get_all_neighbours() if not self.get_wall_between(cell)]

    def get_all_neighbours(self):
        cells = self.maze_generator.cells
        return [self.get_neighbour_cell(i) for i, _ in enumerate(self.neighbours_offsets) if self.get_neighbour_cell(i) is not None]

    def remove_wall_between(self, cell2):
        self.set_wall_between(cell2, False)

    def add_wall_between(self, cell2):
        self.set_wall_between(cell2, True)

    def set_wall_between(self, cell2, value: bool):
        x = self.x - cell2.x
        y = self.y - cell2.y

        if x == 1:
            self.walls['west'] = value
            cell2.walls['east'] = value
        elif x == -1:
            self.walls['east'] = value
            cell2.walls['west'] = value

        if y == 1:
            self.walls['north'] = value
            cell2.walls['south'] = value
        elif y == -1:
            self.walls['south'] = value
            cell2.walls['north'] = value

    def get_wall_between(self, cell2):
        x = self.x - cell2.x
        y = self.y - cell2.y

        if x == 1:
            return self.walls['west']
        elif x == -1:
            return self.walls['east']

        if y == 1:
            return self.walls['north']
        elif y == -1:
            return self.walls['south']

    def get_neighbour_cell(self, offset_index):
        if offset_index < 0 or offset_index > 3:
            raise ValueError("offset must be between 0 and 3")
        offset = self.neighbours_offsets[offset_index]
        x = self.x + offset[0]
        y = self.y + offset[1]

        if x < 0 or x >= self.maze_generator.grid_size or y < 0 or y >= self.maze_generator.grid_size:
            return None

        return self.maze_generator.cells[x][y]
    
    def to_json(self):
        return {
            'x': self.x,
            'y': self.y,
            'walls': self.walls,
            'visited': self.visited,
            'category': self.category.value,
            'route': self.route,
            'distance': self.distance,
            'index': self.index if hasattr(self, 'index') else None,
        }

class MazeGenerator:
    def __init__(self):
        self.cells = []
        self.start_cell = None
        self.end_cell = None
        self.seed = 0
        self.grid_size = 4
        self.sparseness = 0.5
        self.dead_ends_ratio = 0.75
        self.shortcuts_ratio = 0.2
        self.generation_order = EGenerationOrder.Shift
        self.current_cell_index = 0
        self.seeded_random = random.Random(self.seed)
        self.routes = []

    @property
    def max_cells_count(self):
        return self.grid_size ** 2

    def update_generator(self):
        self.seeded_random.seed(self.seed)

    def get_random_cell(self):
        return self.cells[random.randint(0, self.grid_size - 1)][random.randint(0, self.grid_size - 1)]

    def initialize(self):
        self.update_generator()
        self.current_cell_index = 0
        self.cells = [[MazeCell(self, x, y) for y in range(self.grid_size)] for x in range(self.grid_size)]
        self.routes = []

    def generate(self):
        self.initialize()
        stack = []
        current_cell = self.start_cell = self.get_random_cell()
        end_cell = current_cell
        current_cell.distance = 0
        current_cell.index = self.current_cell_index
        current_cell.visited = True
        current_cell.category = ECellCategory.Start
        stack.append(current_cell)
        route = []

        while stack:
            route.append(current_cell)

            if current_cell.index >= max(1, (1 - self.sparseness) * self.max_cells_count):
                if len(route) > 1:
                    self.routes.append(route)
                route = []
                break

            neighbours = current_cell.get_all_neighbours()
            unvisited_neighbours = [cell for cell in neighbours if not cell.visited]

            if unvisited_neighbours:
                random_neighbour = unvisited_neighbours[random.randint(0, len(unvisited_neighbours) - 1)]
                current_cell.remove_wall_between(random_neighbour)
                random_neighbour.visited = True
                random_neighbour.distance = current_cell.distance + 1
                random_neighbour.index = self.current_cell_index + 1
                self.current_cell_index += 1
                stack.append(random_neighbour)
                current_cell = random_neighbour
                end_cell = current_cell if current_cell.distance > end_cell.distance else end_cell
            else:
                if self.generation_order == EGenerationOrder.Pop:
                    current_cell = stack.pop()
                elif self.generation_order == EGenerationOrder.Shift:
                    current_cell = stack.pop(0)

                if len(route) > 1:
                    self.routes.append(route)
                route = []

        end_cell.category = ECellCategory.End
        self.end_cell = end_cell

        # Additional logic for dead ends and shortcuts would be added here, similar to the original TypeScript code

        self.reset_visited()

    def reset_visited(self):
        for row in self.cells:
            for cell in row:
                cell.visited = False

    def to_json(self):
        return {
            'grid_size': self.grid_size,
            'seed': self.seed,
            'sparseness': self.sparseness,
            'dead_ends_ratio': self.dead_ends_ratio,
            'shortcuts_ratio': self.shortcuts_ratio,
            'generation_order': self.generation_order.value,
            'cells': list(map(lambda row: list(map(lambda cell: cell.to_json(), row)), self.cells)),
        }

    # Additional methods like 'to_json' and 'for_each_cell' would be implemented here, similar to the original TypeScript code

def print_pretty_json(data):
    print(json.dumps(data, indent=4))

def print_console_maze(maze_generator: MazeGenerator):
    for y in range(maze_generator.grid_size):
        for x in range(maze_generator.grid_size):
            cell = maze_generator.cells[x][y]
            print(' ', end='')
            print(' ', end='') if cell.walls['north'] else print('_', end='')
        print()

        for x in range(maze_generator.grid_size):
            cell = maze_generator.cells[x][y]
            print('|', end='') if cell.walls['west'] else print(' ', end='')
            print(' ', end='') if cell.category == ECellCategory.Start else print(' ', end='') if cell.category == ECellCategory.End else print(' ', end='')
            print('|', end='') if cell.walls['east'] else print(' ', end='')
        print()

    for x in range(maze_generator.grid_size):
        cell = maze_generator.cells[x][maze_generator.grid_size - 1]
        print(' ', end='')
        print(' ', end='') if cell.walls['south'] else print('_', end='')
    print()


def main():
    generator = MazeGenerator()
    generator.generate()
    print(generator.routes)
    print_console_maze(generator)
    print_pretty_json(generator.to_json())

    # Print or process the generated maze as needed

if __name__ == "__main__":
    main()